File: digital-signage-saas/watchdog-apk/app/src/main/java/com/signage/watchdog/AlarmReceiver.java
package com.signage.watchdog;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.util.Log;

/**
 * Receives the alarm scheduled by BootReceiver and starts WatchdogService.
 */
public class AlarmReceiver extends BroadcastReceiver {
    private static final String TAG = "SignageWatchdog";
    static final String ACTION_START_WATCHDOG = "com.signage.watchdog.START_WATCHDOG";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (ACTION_START_WATCHDOG.equals(intent != null ? intent.getAction() : null)) {
            Log.i(TAG, "Alarm received, starting WatchdogService and launching player");
            launchPlayer(context);
            try {
                Intent serviceIntent = new Intent(context, WatchdogService.class);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    context.startForegroundService(serviceIntent);
                } else {
                    context.startService(serviceIntent);
                }
            } catch (Exception e) {
                Log.e(TAG, "Failed to start WatchdogService", e);
            }
        }
    }

    private void launchPlayer(Context context) {
        try {
            Intent launch = context.getPackageManager().getLaunchIntentForPackage(WatchdogService.TARGET_PACKAGE);
            if (launch == null) {
                launch = new Intent(Intent.ACTION_MAIN);
                launch.setClassName(WatchdogService.TARGET_PACKAGE, WatchdogService.TARGET_PACKAGE + ".MainActivity");
            }
            launch.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP
                | Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED | Intent.FLAG_ACTIVITY_SINGLE_TOP
                | Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
            context.startActivity(launch);
        } catch (Exception e) {
            Log.w(TAG, "Could not launch player at boot: " + e.getMessage());
        }
    }
}

File: digital-signage-saas/signage-unified/app/src/main/java/com/signage/player/AlarmReceiver.kt
package com.signage.player

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build

class AlarmReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent?) {
        if (intent?.action == ACTION_START_WATCHDOG) {
            launchMainActivityOnTop(context)
            val serviceIntent = Intent(context, WatchdogService::class.java).apply {
                putExtra(WatchdogService.EXTRA_SKIP_DELAY, true)
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(serviceIntent)
            } else {
                context.startService(serviceIntent)
            }
        }
    }

    private fun launchMainActivityOnTop(context: Context) {
        val launch = Intent(context, MainActivity::class.java).apply {
            addFlags(
                Intent.FLAG_ACTIVITY_NEW_TASK or
                Intent.FLAG_ACTIVITY_CLEAR_TOP or
                Intent.FLAG_ACTIVITY_REORDER_TO_FRONT or
                Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED or
                Intent.FLAG_ACTIVITY_SINGLE_TOP
            )
        }
        try {
            context.startActivity(launch)
        } catch (_: Exception) { }
    }

    companion object {
        const val ACTION_START_WATCHDOG = "com.signage.unified.START_WATCHDOG"
    }
}
